### PR TITLE
Update testsuite and implement table64

### PR DESCRIPTION
**Pull request by sbc100**
See https://github.com/WebAssembly/memory64/issues/51

Includes a work around for #2388

### CODE REVIEW

**Comment by sbc100** [comment]
PTAL! 

The workaround for #2388 can be replaced by #2389 later I think.

**Comment by sbc100** [comment]
Gentle ping..

File: src/c-writer.cc [hunk 1]
```diff
@@ -1292,7 +1292,14 @@ void CWriter::Write(const Const& const_) {
         // Negative zero. Special-cased so it isn't written as -0 below.
         Writef("-0.0");
       } else {
-        Writef("%.17g", Bitcast<double>(f64_bits));
+        char buf[128];
+        snprintf(buf, sizeof(buf), "%.17g", Bitcast<double>(f64_bits));
+        // Append .0 to if sprint didn't include a decimal point.
+        // Works around https://github.com/WebAssembly/wabt/issues/2422
+        if (!strchr(buf, '.') && !strchr(buf, '+')) {
```
Comment (SoniEx2) [inline src/c-writer.cc:1299]: `'+'`?

Comment (sbc100) [reply src/c-writer.cc:1299]: This avoids adding `.0` to constants that look like `1.2200011354000000e+28`

Comment (SoniEx2) [reply src/c-writer.cc:1299]: but that has a `.` already, do you mean like `1e-100`?

**Comment by keithw** [comment]
I am totally swamped right now but will get to this asap (probably Friday evening). Sorry for delay.

File: src/c-writer.cc (hunk 1 above)
Comment (sbc100) [reply src/c-writer.cc]: Updated to look for `e` instead, and added a comment.

File: src/c-writer.cc [hunk 2]
```diff
@@ -1292,7 +1292,15 @@ void CWriter::Write(const Const& const_) {
         // Negative zero. Special-cased so it isn't written as -0 below.
         Writef("-0.0");
       } else {
-        Writef("%.17g", Bitcast<double>(f64_bits));
+        char buf[128];
+        snprintf(buf, sizeof(buf), "%.17g", Bitcast<double>(f64_bits));
+        // Append .0 to if sprint didn't include a decimal point or use the
```
Comment (dschuff) [inline src/c-writer.cc:1297]:
```suggestion
        // Append .0 to buf if sprint didn't include a decimal point or use the
```

**Review by dschuff — APPROVED** [review]
can you update the commit message to also cover what's happening with the tests (e.g. EH etc)

**Comment by sbc100** [comment]
> can you update the commit message to also cover what's happening with the tests (e.g. EH etc)

Oops, good catch I didn't actually mean to delete those files.  I took care of that already in #2416

File: include/wabt/type-checker.h [hunk 3]
```diff
@@ -77,7 +77,8 @@ class TypeChecker {
   Result EndBrTable();
   Result OnCall(const TypeVector& param_types, const TypeVector& result_types);
   Result OnCallIndirect(const TypeVector& param_types,
-                        const TypeVector& result_types);
+                        const TypeVector& result_types,
+                        const Limits& table_limits);
```
Comment (dimitry-) [inline include/wabt/type-checker.h:81]: does OnReturnCallIndirect need the same modification?

Comment (sbc100) [reply include/wabt/type-checker.h:81]: Most likely.  I guess the test suite is lacking there?

**Pull request merged**